src/amici.cpp

```diff
@@ -1975,88 +1895,339 @@ rdata->s ## QUANT [(ip*N1 + N1T + IND1)*N2 + IND2] += udata->k[udata->nk-udata->
     }
     
     return status;
 }
 
-int fsJy(realtype t, int it, realtype *sJy, realtype *s2Jy, realtype *dJydy, realtype *dJydp, realtype *y, realtype *sigmay, realtype *sy, realtype *dydp, realtype *my, void *user_data) {
+static int fsJy(int it, UserData *udata, TempData *tdata, const ExpData *edata, ReturnData *rdata) {
+    int status = AMICI_SUCCESS;
+    
+    // Compute dJydx*sx for current 'it'
+    // dJydx        nt x nJ x nx
+    // sx           nt x nx x nplist
+    
+    realtype *multResult = new double[udata->nJ * udata->nplist]();
+    realtype *dJydxTmp = new double[udata->nJ * udata->nx];
+    realtype *sxTmp = new double[udata->nplist * udata->nx];
+    for(int ix = 0; ix < udata->nx; ++ix){
+        for(int ip = 0; ip < udata->nplist; ++ip)
+            sxTmp[ix + ip * udata->nx] = rdata->sx[it + (ix + ip * udata->nx ) * udata->nt];
+        for(int iJ = 0; iJ < udata->nJ; ++iJ)
+            dJydxTmp[iJ + ix * udata->nJ] = tdata->dJydx[it + (iJ + ix * udata->nJ ) * udata->nt];
+    }
+    
+    // C := alpha*op(A)*op(B) + beta*C,
+    amici_dgemm(AMICI_BLAS_ColMajor, AMICI_BLAS_NoTrans, AMICI_BLAS_NoTrans,
+                udata->nJ, udata->nplist, udata->nx,
+                1.0, dJydxTmp, udata->nJ,
+                sxTmp, udata->nx,
+                1.0, multResult, udata->nJ);
+    
+    // multResult    nJ x nplist
+    // dJydp         nJ x nplist
+    // dJydxTmp      nJ x nx
+    // sxTmp         nx x nplist
+    
+    // sJy += multResult + dJydp
+    for(int iJ = 0; iJ < udata->nJ; ++iJ) {
+        if (iJ == 0)
+            for(int ip = 0; ip < udata->nplist; ++ip)
+                rdata->sllh[ip] -= multResult[ip * udata->nJ] + tdata->dJydp[ip * udata->nJ];
+        else
+            for(int ip = 0; ip < udata->nplist; ++ip)
+                rdata->s2llh[(iJ - 1) + ip * (udata->nJ-1) ] -= multResult[iJ + ip * udata->nJ] + tdata->dJydp[iJ + ip * udata->nJ];
+    }
+    
+    delete[] dJydxTmp;
+    delete[] multResult;
+    delete[] sxTmp;
+    
+    return(status);
+}
+
+int fdJydp(int it, UserData *udata, TempData *tdata, const ExpData *edata, ReturnData *rdata) {
+    int status = AMICI_SUCCESS;
+    
+    // dJydy         nytrue x nJ x ny
+    // dydp          ny x nplist
+    // dJydp         nJ x nplist
+    
+    memset(tdata->dJydp,0,udata->nJ * udata->nplist * sizeof(double));
+    
+    realtype *dJydyTmp = new double[udata->nJ * udata->ny];
+    realtype *dJydsigmaTmp = new double[udata->nJ * udata->ny];
+    
+    for(int iyt=0; iyt < udata->nytrue; ++iyt) {
+        if (amiIsNaN(edata->my[udata->nt * iyt + it]))
+            continue;
+        
+        // copy current (iyt) dJydy and dJydsigma slices
+        // dJydyTmp     nJ x ny
+        // dJydsigmaTmp nJ x ny
+        for(int iJ = 0; iJ < udata->nJ; ++iJ) {
+            for(int iy = 0; iy < udata->ny; ++iy) {
+                dJydyTmp[iJ + iy * udata->nJ] = tdata->dJydy[iyt + (iJ + iy * udata->nJ) * udata->nytrue];
```

Comment:
We have these non-contiguous vectors in so many places. Is this due to how we get data from cvodes or due to matlab col-major storage? Is there a chance to change the memory layout and avoiding all the extra copying or do we anyways need to slice in all directions?